python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -740,3 +742,111 @@ def with_reconstruction_metadata(self, state_metadata: Mapping[str, Any]) -> Sel
                 **state_metadata,
             },
         )
+
+    @public
+    def map_asset_specs(self, fn: Callable[[AssetSpec], AssetSpec]) -> "Definitions":
+        """Applies the given mapping function to every asset spec within this Definitions object,
+        including both external asset specs and asset specs within AssetsDefinitions.
+
+        Returns a new Definitions object with the transformed assets; does not mutate this object.
+
+        If the mapping function produces asset specs with new asset keys, downstream assets and
+        asset checks will be updated so that they depend on the new key.
+
+        Does not support Definitions  objects that contain CacheableAssetsDefinitions.
+
+        Args:
+            fn (Callable[[AssetSpec], AssetSpec]): A function that accepts an AssetSpec and returns
+                a new AssetSpec.
+
+        Returns:
+            Definitions: A new Definitions object with the transformed assets.
+        """
+        if self.assets is None:
+            return self
+
+        new_keys_by_old_key: Dict[AssetKey, AssetKey] = {}
+
+        assets_def_new_specs: List[Tuple[AssetsDefinition, Sequence[AssetSpec]]] = []
+        new_external_asset_specs: List[AssetSpec] = []
+        new_source_assets: List[SourceAsset] = []
+
+        for el in self.assets:
```

Comment:
The fact that we don't canonicalize these to assets definitions makes this very complicated. @dpeng817 and I are encountering this friction a lot in our airlift work as well.

Reply:
Yeah I think this is a generalized problem with `Definitions` right now as an `assets` property seems very first-class.

Reply:
Yeah that makes sense. 

A couple thoughts:

* I think the rule should be that the moment you are processing the definitions meaningfully they should be canonicalized to their underlying form.
* I think we should stash the input parameters somewhere more hidden and use them only internally (for merge only for now).